### PR TITLE
Core: Integrate serverChannel into channel

### PR DESCRIPTION
**Pull request by ndelangen**
Closes #

Add listening capability to serverChannel on node's side
Fix data/args handling from serverChannel

## What I did

- addons can now extend the serverChannel from `preset.ts`
- addons can now listen to channel messages and messages emitted from the serverChannel will appear
- addons can now emit to the channel and messages will also be send to the server
- Channel class can now receive multiple transports
- add deprecation notices to serverChannel everywhere the channel should be used instead

## How to test

add this to a `manager.ts`:
```ts
import * as React from 'react';
import { addons, types } from '@storybook/manager-api';

addons.register('ADDON_ID', (api) => {
  addons.add('ADDON_ID', {
    title: 'debugger',
    id: 'debugger',
    type: types.PANEL,
    match: ({ viewMode }) => viewMode === 'story',
    render: () => {
      const [state, setState] = React.useState('...');
      React.useEffect(() => {
        api.on('storybook/debug-b', (data) => {
          setState(JSON.stringify(data));
        });
      }, []);
      return React.createElement(
        'div',
        null,
        React.createElement('pre', null, state),
        React.createElement(
          'button',
          {
            type: 'button',
            onClick: () => {
              api.emit('storybook/debug-a', { foo: 'bar' });
            },
          },
          'button'
        )
      );
    },
  });
});
```

add this to `preset.ts`:
```ts
export const experimental_serverChannel = (channel: Channel) => {
  channel.on('storybook/debug-a', (event) => {
    console.log(event);
    channel.emit('storybook/debug-b', event);
  });
  return channel;
};
```

1. opening the storybook and go to any (noon docs) story
2. open the debugger panel
3. click the button in the debugger panel
4. notice how the CLI prints some data
5. notice how the `pre`-tag gets filled with data.
6. you can observe the websocket in the network tab

The channel should work correctly as well in build-mode

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] If this PR should be tested against many or all sandboxes,
      make sure to add the `ci:merged` or `ci:daily` GH label to it.
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

**Comment by tmeasday** [comment]
@ndelangen I made some comments on https://github.com/storybookjs/storybook/pull/22939

File: code/lib/core-server/src/utils/get-server-channel.ts [hunk 1]
```diff
@@ -1,32 +1,52 @@
...
-import { stringify } from 'telejson';
+import { isJSON, parse, stringify } from 'telejson';
+import type { ChannelHandler } from '@storybook/channels';
+import { Channel } from '@storybook/channels';
 
 type Server = ConstructorParameters<typeof WebSocketServer>[0]['server'];
 
-export class ServerChannel {
-  webSocketServer: WebSocketServer;
+export class ServerChannelTransport {
```
Comment (tmeasday) [inline code/lib/core-server/src/utils/get-server-channel.ts:8]: Can we add a comment explaining that unlike other channels, this channel can connect to >1 client and a message sent out on the server channel can reach multiple (browser) endpoints?

File: code/lib/core-server/src/dev-server.ts [hunk 2]
```diff
@@ -29,7 +29,7 @@ export async function storybookDevServer(options: Options) {
     options.presets.apply<CoreConfig>('core'),
   ]);
 
-  const serverChannel = getServerChannel(server);
+  const serverChannel = await options.presets.apply('serverChannel', getServerChannel(server));
```
Comment (tmeasday) [inline code/lib/core-server/src/dev-server.ts:32]: Should we call this `experiemental_xyz` for now?

Also, the name `serverChannel` is a little bit unintuitive to me. That implies I'd be providing rather than modifying a server channel. I'm not sure of a better name though.

**Review by tmeasday — CHANGES_REQUESTED** [review]
Would like to consider creating a new kind of channel too, so postmessage channel doesn't implicitly contain a websocket channel (or just renaming postmessage channel). But how will this multiplexing work in RN where both channels are WS?

**Comment by ndelangen** [comment]
> But how will this multiplexing work in RN where both channels are WS?

I should probably make the same multiplex change to the websocket channel package?
I hadn't really considered RN.

File: code/lib/core-server/src/dev-server.ts (hunk 2 above)
Comment (ndelangen) [reply code/lib/core-server/src/dev-server.ts:32]: Any preset value gets the default value, and gets the option to do something with it, modify it, return something new.

I don't have a better name either..

**Comment by ndelangen** [comment]
@valentinpalkovic I wasn't planning to.

File: code/ui/blocks/src/blocks/external/ExternalPreview.ts [hunk 3]
```diff
@@ -36,7 +36,7 @@ export class ExternalPreview<TRenderer extends Renderer = Renderer> extends Prev
   private moduleExportsByImportPath: Record<Path, ModuleExports> = {};
 
   constructor(public projectAnnotations: ProjectAnnotations<TRenderer>) {
-    super(new Channel());
+    super(new Channel({}));
```
Comment (yannbf) [inline code/ui/blocks/src/blocks/external/ExternalPreview.ts:39]: is this change needed? seems like the constructor has a default value if not passed? I'm asking because if it is needed, then this seems technically a breaking change

**Comment by yannbf** [comment]
This looks amazing! I wonder if there's some documentation we can add to this?

File: code/ui/blocks/src/blocks/external/ExternalPreview.ts (hunk 3 above)
Comment (ndelangen) [reply code/ui/blocks/src/blocks/external/ExternalPreview.ts:39]: runtime? no.
I made the type stricter though.

Creating a Channel without a transport is completely pointless, who creates channels other then us in the core?

**Comment by ndelangen** [comment]
@yannbf the idea is that it's experimental. We'll try to do a few things internally, test if we like it and maybe iterate on it before announcing it to the public for general use.

Documenting it would certainly be welcome. How would you recommend we describe this change, in what document?

File: code/ui/blocks/src/blocks/external/ExternalPreview.ts (hunk 3 above)
Comment (yannbf) [reply code/ui/blocks/src/blocks/external/ExternalPreview.ts:39]: got it, and I don't know, so I thought I'd flag this, on the safe side. Feel free to ignore this then!

**Review by yannbf — APPROVED** [review]
LGTM! It's working great and fulfills the needs of the onboarding 👯

**Comment by tmeasday** [comment]
> I should probably make the same multiplex change to the websocket channel package?
> I hadn't really considered RN.

Maybe we should wait until we actually implement it in RN but I'm wondering if instead of the current implementation where `@storybook/channel-X` actually creates two channels (not even both of type X), we instead should add a new API to the `@storybook/channel` package where it can take two transports, and just do something like:

```
// in manager / preview code
import { websocketTransport } from '@storybook/channel-websocket';
import { postmessageTransport } from '@storybook/channel-websocket';

this.channel = new Channel({ transports: [websocketTransport, postmessageTransport] });
```

**Comment by ndelangen** [comment]
I really don't like the idea of creating yet-another package, that is counter to what I'm actually trying to achieve long term.

I don't agree that moving the code into each builder (twice because of SSv6 support) is the right move, but I'll do it.

I've also deprecated the `createChannel` function on `channel-postmessage` (but NOT on `channel-websocket` (in fact I completely reverted `channel-websocket`'s code.
`createChannel` of `postMessage` is to be reverted as well I guess, nothing should be using it anymore anyway.

**Comment by shilman** [comment]
@tmeasday discussed with @ndelangen and he's going to revert the last set of changes and merge to unblock @yannbf . Let's spend a few minutes together tomorrow to discuss the ideal API here since the changes you suggested above result in a bunch of code duplication in the builders & @ndelangen doesn't want to introduce a new package. He has a better idea -- we think -- but would be good to hammer it out together synchronously.

**Comment by ndelangen** [comment]
@shilman I made a spike of moving the code to lib/channels, here: https://github.com/storybookjs/storybook/pull/23032
